### PR TITLE
LPS-161163 Add Version History to the Info Panel for Documents and Media

### PR DESCRIPTION
**Pull request by jkappler**
[Jira story ticket](https://issues.liferay.com/browse/LPS-160208)

## Motivation

There are cases were Content Marketers want to see a Document's version history in the Info Panel of the Content Dashboard, so that they can easily understand the history of that document.

## Proposed solution

The idea would be to change the UI of the side bar so that instead of having collapsible panels we would have 2 tab panes, one for details, and the other one for the version of the items

## Gif of how to test it

1. Enable Feature Flag in portal-ext.properties: feature.flag.LPS-161013=true (LPS of the epic)
2. Create a new document and edit it several times so that we have some history
3. Go to Content Dashboard and click the info icon to check the details
4. Now instead of a collapsible section we'll see 2 tabs, one for details and the other for versions, just as the gif

versions_history

### CODE REVIEW

**Comment by jkappler** [comment]
@cgoncas, I adapted the UI, but it will surely require more ❤️.

As for the Version class, I decided not to move to Info Fw at the moment. And I'm unsure of the orders in the constructor, maybe have it ordered? I left it in order of importance

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 1]
```diff
@@ -218,6 +224,35 @@ protected void doServeResource(
...
+		ContentDashboardItem contentDashboardItem, ThemeDisplay themeDisplay) {
+
+		int status = WorkflowConstants.STATUS_APPROVED;
+
+		PermissionChecker permissionChecker =
+			themeDisplay.getPermissionChecker();
+
+		User user = themeDisplay.getUser();
+
+		if ((user.getUserId() == contentDashboardItem.getUserId()) ||
```
Comment (liferay-tango) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:237]: Hey @jkappler !!

What do you think about moving that logic to the ContentDashboardItem, so every implementation can decide what are the status of the versions they should return?

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/item/FileEntryContentDashboardItem.java [hunk 2]
```diff
@@ -107,6 +107,26 @@ public FileEntryContentDashboardItem(
 		_portal = portal;
 	}
 
+	@Override
+	public List<Version> getAllVersions(Locale locale, int status) {
+		Stream<FileVersion> stream = _fileEntry.getFileVersions(
```
Comment (liferay-tango) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/item/FileEntryContentDashboardItem.java:112]: Hey @jkappler !!

What do you think about limit the versions results? It is possible to break the frontend if we return too many versions?

Thanks!!

**Comment by liferay-tango** [comment]
Hey @Stephy6 !!

Could you take a look to the frontend part?

Thanks!!!

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java (hunk 1 above)
Comment (jkappler) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:237]: Hi @cgoncas, will change this

**Comment by Stephy6** [comment]
> Hey @Stephy6 !!
> 
> Could you take a look to the frontend part?
> 
> Thanks!!!

Yes sure!

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:tango

**Comment by jkappler** [comment]
will check the issues

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java (hunk 1 above)
Comment (jkappler) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java]: Updated PR with the changes, every ContentDashboardItem will decided what to return

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/item/FileEntryContentDashboardItem.java (hunk 2 above)
Comment (jkappler) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/item/FileEntryContentDashboardItem.java:128]: Will be handled in a new story

**Comment by jkappler** [comment]
@Stephy6, I rebased the PR and fixed the js tests

**Comment by jkappler** [comment]
ci:test:relevant

**Comment by jkappler** [comment]
ci:test:tango

**Comment by jkappler** [comment]
ci:test:sf

**Comment by jkappler** [comment]
fixed sf issue

**Comment by jkappler** [comment]
ci:test:relevant

**Comment by jkappler** [comment]
ci:test:tango

**Comment by jkappler** [comment]
ci:test:relevant

**Comment by jkappler** [comment]
ci:test:tango

**Review by cgoncas — APPROVED** [review]
For the backend part I think the PR is good.

**Comment by jkappler** [comment]
ci:test:sf

**Comment by jkappler** [comment]
conflicts resolved

**Comment by beltranrengifo** [comment]
hey @jkappler @Stephy6  !
I've just pushed the changes we did this morning. Also, I couldn't resist to move all the taxonomies block to its own component. Plz take a look whenever you can and let me know. Thxs!

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
@jkappler I'm not sure why the sf is failing... The command returns an OK in my local installation

**Comment by beltranrengifo** [comment]
Seems something related to Typescript, some types not being generated. I'm trying to build the portal. I'll let you know in a while!

**Comment by jkappler** [comment]
This is weird, can't reproduce this in local, with gw or ant

**Comment by jkappler** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by jkappler** [comment]
ci:forward

**Comment by beltranrengifo** [comment]
> ci:forward

I'm afraid this probably will not pass the CI, the FE units are broken due to props naming/structure changes I believe... :(
I'll try to figure this out tomorrow, along with adding more test for the new components, if needed... 🙌

**Comment by jkappler** [comment]
ci:stop

**Comment by jkappler** [comment]
Stopping it for now @beltranrengifo, I rushed too much 🗡️

**Comment by beltranrengifo** [comment]
Test are green! Running the CI again @jkappler

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:relevant

**Comment by beltranrengifo** [comment]
ci:reevaluate:1357401_2456

**Comment by beltranrengifo** [comment]
ci:test:relevant

**Comment by jkappler** [comment]
ci:test:relevant

**Comment by jkappler** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
Looks like there are some test failing in [.ContentDashboardAdminPortletTest.testGetSearchContainerWithStatusDraftAndHasApprovedVersion](https://test-1-27.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/196746//testReport/com.liferay.content.dashboard.web.internal.portlet.test/ContentDashboardAdminPortletTest/testGetSearchContainerWithStatusDraftAndHasApprovedVersion), I think they are related with this PR.

I don't know why they are marked as passing :(


[com.liferay.content.dashboard.web.internal.portlet.test.ContentDashboardAdminPortletTest.testGetSearchContainerWithStatusDraftAndHasApprovedVersion](https://test-1-27.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/196746//testReport/com.liferay.content.dashboard.web.internal.portlet.test/ContentDashboardAdminPortletTest/testGetSearchContainerWithStatusDraftAndHasApprovedVersion)
java.lang.NoSuchMethodException: No method on null with name getVersions and parameter types [class java.util.Locale]

  at com.liferay.portal.kernel.test.ReflectionTestUtil.getMethod(ReflectionTestUtil.java:190)

  at com.liferay.portal.kernel.test.ReflectionTestUtil.invoke(ReflectionTestUtil.java:217)

  at com.liferay.content.dashboard.web.internal.portlet.test.ContentDashboardAdminPortletTest.testGetSearchContainerWithStatusDraftAndHasApprovedVersion(ContentDashboardAdminPortletTest.java:1390)

  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

  at java.lang.reflect.Method.invoke(Method.java:498)

  at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable$3.evaluate(TestExecutorRunnable.java:353)

  at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)

  at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)

  at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)

  at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)

  at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)

  at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)

  at com.liferay.portal.kernel.test.rule...

**Comment by cgoncas** [comment]
> @jkappler @beltranrengifo
> 
> Looks like there are some tests failing in [.ContentDashboardAdminPortletTest.testGetSearchContainerWithStatusDraftAndHasApprovedVersion](https://test-1-27.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/196746//testReport/com.liferay.content.dashboard.web.internal.portlet.test/ContentDashboardAdminPortletTest/testGetSearchContainerWithStatusDraftAndHasApprovedVersion), I think they are related with this PR.
> 
> I don't know why they are marked as passing :(
> 
> [com.liferay.content.dashboard.web.internal.portlet.test.ContentDashboardAdminPortletTest.testGetSearchContainerWithStatusDraftAndHasApprovedVersion](https://test-1-27.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/196746//testReport/com.liferay.content.dashboard.web.internal.portlet.test/ContentDashboardAdminPortletTest/testGetSearchContainerWithStatusDraftAndHasApprovedVersion) java.lang.NoSuchMethodException: No method on null with name getVersions and parameter types [class java.util.Locale]
> 
> at com.liferay.portal.kernel.test.ReflectionTestUtil.getMethod(ReflectionTestUtil.java:190)
> 
> at com.liferay.portal.kernel.test.ReflectionTestUtil.invoke(ReflectionTestUtil.java:217)
> 
> at com.liferay.content.dashboard.web.internal.portlet.test.ContentDashboardAdminPortletTest.testGetSearchContainerWithStatusDraftAndHasApprovedVersion(ContentDashboardAdminPortletTest.java:1390)
> 
> at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 
> at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 
> at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 
> at java.lang.reflect.Method.invoke(Method.java:498)
> 
> at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable$3.evaluate(TestExecutorRunnable.java:353)
> 
> at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
> 
> at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
> 
> at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
> 
> at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
> 
> at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
> 
> at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
> 
> at com.liferay.portal.kernel.test.rule...

I wil try to fix it @beltranrengifo and @jkappler

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:sf

**Pull request closed without merging**